File: typescript/queue/queue.ts

class Node {
    constructor(public value: any, public next?: Node) { }
}

export class Queue {
    private front: Node | null = null
    private rear: Node | null = null
    private length: number = 0
    public enqueue(value: any) {
        const node = new Node(value)
        if (!this.front) {
            this.front = node
        }
        if (this.rear) {
            this.rear.next = node
        }
        this.rear = node
        this.length++
    }

    public dequeue() {
        if (!this.front) return
        const node = this.front
        this.front = node.next!
        if (!this.front) {
            this.rear = null
        }
        this.length--
        return node.value
    }

    isEmpty() {
        return !this.front
    }

    size() {
        return this.length
    }
}

File: Java/src/main/java/com/example/queue/Queue.java
package com.example.queue ;

class Node<T> {
    T data;
    Node<T> next;

    public Node(T data) {
        this.data = data;
        next = null;
    }
}

public class Queue<T> {

    private Node<T> front;
    private Node<T> rear;
    private int length;

    public Queue() {
        front = null;
        rear = null;
        length = 0;
    }

    public void enqueue(T data) {
        Node<T> newNode = new Node<T>(data);
        if (front == null) {
            front = newNode;
        } else {
            rear.next = newNode;
        }
        rear = newNode;
        length++;
    }

    public T dequeue() {
        if (front == null) {
            return null;
        }
        T data = front.data;
        front = front.next;
        if (front == null) {
            rear = null;
        }
        length--;

        return data;
    }

    public T peek() {
        if (front == null) {
            return null;
        }
        return front.data;
    }

    public int size() {
        return length;
    }

    public boolean isEmpty() {
        return front == null;
    }
}